network_analysis/trace-a-utility-network/README.md

```diff
@@ -0,0 +1,54 @@
+# Trace a utility network
+
+Discover connected features in a utility network using connected, subnetwork, upstream, and downstream traces.
+
+![](TraceUtilityNetwork.png)
+
+## Use case
+
+You can use a trace to visualize and validate the network topology of a utility network for quality assurance. Subnetwork traces are used for validating whether subnetworks, such as circuits or zones, are defined or edited appropriately.
+
+## How to use the sample
+
+Click on one or more features while 'Add starting locations' or 'Add barriers' is selected. When a junction feature is identified, you may be prompted to select a terminal to use as the starting location/barrier. When an edge feature is identified, the distance from the tapped location to the beginning of the edge feature will be computed. Select the type of trace using the drop down menu. Click 'Trace' to initiate a trace on the network and see the results selected. Click 'Reset' to clear the trace parameters and start over.
+
+## How it works
+
+1. Create a `Map` and add it to a `MapView`.
+2. Using the URL to a utility network's feature service, create `FeatureLayer`s that contain the utility network's features, and add them to the operational layers of the map.
+3. Create and load a `UtilityNetwork` with the same feature service URL and map.
+4. Add a `GraphicsOverlay` with symbology that distinguishes starting points from barriers.
+5. Add a listener for clicks on the map view, and use `mapView.identifyLayersAsync()` to identify clicked features.
+6. Create a `UtilityElement` that represents its purpose (starting point or barrier) at the location of each identified feature, and save it to a list.
+7. Determine the type of the identified feature using `utilityNetwork.getDefinition().getNetworkSource()` passing its table name.
+8. If a junction, display a terminal picker when more than one `UtilityTerminal` is found and create a `UtilityElement` with the selected terminal or the single terminal if there is only one.
+9. If an edge, create a utility element from the identified feature and set its `FractionAlongEdge` using `GeometryEngine.fractionAlong()`.
+10. Create `TraceParameters` with the selected trace type along with the collected starting locations and barriers (if applicable).
+11. Set the `TraceParameters.TraceConfiguration` with the utility tier's `TraceConfiguration` property.
+12. Run a `utilityNetwork.traceAsync()` with the specified parameters, and get the `UtilityTraceResult`.
+13. From the utility trace result, get the `UtilityElementTraceResult`, and group the utility elements within by their network source name, which can be retrieved using `utilityElement.getNetworkSource().getName()`.
+13. For every feature layer in this map with elements, select features by converting utility elements to `ArcGISFeature`(s) using `UtilityNetwork.fetchFeaturesForElementsAsync()`
```

Comment:
I can't follow this in the code, not seeing a `fetchFeaturesForElementsAsync()` method?

Reply:
This was lost in translation from .NET, and not up to date anyway. We've used a different approach in Java so that the UI is disabled until the selection is done. Updated to reflect what's going on in the code.